### PR TITLE
Add type for objects where undefined is not allowed

### PR DESCRIPTION
**Pull request by cailyncodes**
There was several places where calling code needed to be updated, because typescript wasn't powerful enough for some of these type inferences.

See commit message: Enrollment flow is broken.

Here is the db error: `The INSERT statement conflicted with the FOREIGN KEY constraint "FK_Family_Organization_OrganizationId". The conflict occurred in database "master", table "dbo.Organization", column 'Id'.`

### CODE REVIEW

File: src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx [hunk 1]
```diff
@@ -33,9 +34,9 @@ export default function ReportDetail() {
     return <div className="Report"></div>;
   }
 
-  const reportEnrollments = idx(report, _ => _.organization.sites[0].enrollments);
+  const reportEnrollments = report.organization && report.organization.sites && report.organization.sites[0].enrollments;
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx]: why this change?

File: src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx [hunk 2]
```diff
@@ -33,9 +34,9 @@ export default function ReportDetail() {
...
   let numEnrollmentsMissingInfo = reportEnrollments
-    ? reportEnrollments.filter(enrollment => missingInformation(enrollment as Enrollment)).length
+    ? reportEnrollments.filter<DeepNonUndefineable<Enrollment>>((enrollment => missingInformation(enrollment)) as (_: DeepNonUndefineable<Enrollment>) => _ is DeepNonUndefineable<Enrollment>).length
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx]: `:exploding_head:`

File: src/Hedwig/ClientApp/src/containers/Reports/Reports.tsx [hunk 3]
```diff
@@ -82,10 +87,10 @@ export default function Reports() {
     ],
   };
 
-  const submittedTableProps: TableProps<Required<Report>> = {
+  const submittedTableProps: TableProps<DeepNonUndefineable<Report>> = {
     ...defaultTableProps,
     id: 'submitted-reports-table',
-    data: reports.filter(report => !!report.submittedAt),
+    data: [], //reports.filter(report => !!report.submittedAt),
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/containers/Reports/Reports.tsx]: left over debug?

File: src/Hedwig/ClientApp/src/utils/types.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+type FunctionWithRequiredReturnType<
+  T extends (...args: any[]) => any
+> = T extends (...args: infer A) => infer R
+  ? (...args: A) => DeepNonUndefineable<R>
+	: never;
+
+export type DeepNonUndefineable<T> = 
+	T extends any[] ? DeepNonUndefineableArray<T[number]> :
+	T extends (...args: any[]) => any ? FunctionWithRequiredReturnType<T> :
+	T extends object ? DeepNonUndefineableObject<T> : NonUndefineable<T>;
```
Comment (jlhogan) [inline src/Hedwig/ClientApp/src/utils/types.ts]: all of this is wiiiiiild. so cool ! and a lil confusing/intimidating. At some point comments to make more accessible to code readers would rock.

File: src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx (hunk 1 above)
Comment (cailyncodes) [reply src/Hedwig/ClientApp/src/containers/ReportDetail/ReportDetail.tsx]: Idx returns a wrapper type so I changed it so we didn't run into that issue but now that you point this out, I think I can recast it to avoid the issue.

File: src/Hedwig/ClientApp/src/containers/Reports/Reports.tsx (hunk 3 above)
Comment (cailyncodes) [reply src/Hedwig/ClientApp/src/containers/Reports/Reports.tsx]: oops yep - thanks for the catch

File: src/Hedwig/ClientApp/src/utils/types.ts (hunk 4 above)
Comment (cailyncodes) [reply src/Hedwig/ClientApp/src/utils/types.ts]: 💯 Comments coming soon!

**Comment by cailyncodes** [comment]
Just for the record, I'm (symbolically) approving @jlhogan's backend changes on this PR
